rust/gui-client/src-tauri/src/client/gui.rs

```diff
@@ -901,17 +969,31 @@ async fn run_controller(
         tokio::select! {
             result = network_notifier.notified() => {
                 result?;
-                if controller.status.connlib_is_up() {
-                    tracing::debug!("Internet up/down changed, calling `Session::reconnect`");
-                    controller.ipc_client.reset().await?;
+                match &controller.status {
+                    Status::Disconnected => {}
+                    Status::TunnelReady { .. } | Status::WaitingForPortal { .. } | Status::WaitingForTunnel { .. } => {
+                        tracing::debug!("Internet up/down changed, calling `Session::reconnect`");
+                        controller.ipc_client.reset().await?
+                    },
+                    Status::WaitingForNetwork { token } => {
+                        tracing::debug!("Internet up/down changed, retrying Portal connection...");
+                        controller.start_session(token.expose_secret().clone().into()).await?;
+                    },
                 }
             },
             result = dns_notifier.notified() => {
                 result?;
-                if controller.status.connlib_is_up() {
-                    let resolvers = firezone_headless_client::dns_control::system_resolvers_for_gui()?;
+                match &controller.status {
+                    Status::Disconnected => {}
+                    Status::TunnelReady { .. } | Status::WaitingForPortal { .. } | Status::WaitingForTunnel { .. } => {
+                        let resolvers = firezone_headless_client::dns_control::system_resolvers_for_gui()?;
                     tracing::debug!(?resolvers, "New DNS resolvers, calling `Session::set_dns`");
                     controller.ipc_client.set_dns(resolvers).await?;
+                    },
+                    Status::WaitingForNetwork { token } => {
+                        tracing::debug!("New DNS resolvers, retrying Portal connection...");
```

Comment:
Does this effectively mean we trigger on DNS changes? I suppose on Windows that's pretty reliable, but what's the behavior on Linux?

Reply:
It should be DNS changes and network changes on both OSes
- Linux DNS via D-Bus https://github.com/firezone/firezone/blob/4ec78f70c220f564e143ed4667543cd58a2baa6b/rust/bin-shared/src/network_changes/linux.rs#L40
- Linux network via D-Bus https://github.com/firezone/firezone/blob/4ec78f70c220f564e143ed4667543cd58a2baa6b/rust/bin-shared/src/network_changes/linux.rs#L61
- Windows DNS via registry listener https://github.com/firezone/firezone/blob/4ec78f70c220f564e143ed4667543cd58a2baa6b/rust/bin-shared/src/network_changes/windows.rs#L381
- Windows network via complicated COM callback https://github.com/firezone/firezone/blob/4ec78f70c220f564e143ed4667543cd58a2baa6b/rust/bin-shared/src/network_changes/windows.rs#L122thing

Reply:
The only alternative I could think of was polling which didn't sound pleasant